src/Type.h

```diff
@@ -142,20 +142,46 @@ template<> struct halide_c_type_to_name<float> { static const bool known_type =
 template<> struct halide_c_type_to_name<double> { static const bool known_type = true; static halide_cplusplus_type_name name() { return { halide_cplusplus_type_name::Simple,  "double"}; } };
 template<> struct halide_c_type_to_name<struct buffer_t> { static const bool known_type = true; static halide_cplusplus_type_name name() { return { halide_cplusplus_type_name::Struct,  "buffer_t"}; } };
 
+// Default case (should be only Unknown types, since we specialize for Known types below).
+// We require that all unknown types be pointers, and translate them all to void*
+// (preserving const-ness and volatile-ness).
 template<typename T, bool KnownType>
 struct halide_internal_handle_traits {
-    static const halide_handle_cplusplus_type *type_info(bool is_ptr, halide_handle_cplusplus_type::ReferenceType ref_type) { return nullptr; }
+    static const halide_handle_cplusplus_type *type_info(bool is_ptr, 
+            halide_handle_cplusplus_type::ReferenceType ref_type) { 
+        static_assert(!KnownType, "Only unknown types handled here");
+        internal_assert(is_ptr) << "Unknown types must be pointers";
+        internal_assert(ref_type == halide_handle_cplusplus_type::NotReference) << "Unknown types must not be references";
+        static const halide_handle_cplusplus_type the_info{
```

Comment:
Edit, after refreshing my memory on the IRNode issue:
(1) Yes, this looks like it is subject to exactly the same problem. Why it's not already biting us is not clear; perhaps we're just getting lucky, or perhaps this code is under-exercised in our tests.
(2) The solution that Andrew used for IRNode won't really work here, since it requires having an explicit enum with all possible values spelled out; we need something open-ended for unknown and extensible types.

I'll ponder this.